S12 devel

jjgomezcadenas opened the pull request:
This is a major re-factoring of code related with PMAPS. The main changes are:

1) high level objects have been defined for S1, S2 and S2Si. These are classes with persistent capabilities that provide a high level interface to the underlying objects hitherto called by the same name. 

2) The notation has been clarified, so that now: s1, s2 and s2si are instances of classes S1, S2 and S2Si (we can refer to any of these high level objects as si). Instead,  s1d, s2d and s2sid are the underlying dictionaries previously used to represent the si. Thus, a s1d is  a {peak_no: [t, E]}, where t and E are np. arrays. The sid are still accessible through the si (e.g, s1.s1d and so on). However the si provide a better interface, si.peaks where peaks = {peak_no: Peak(t,E)} and Peak itself is a class that provides a high level interface to the peak waveform. 

3) When PMAPS are read (e.g, function load_pmaps) the objects returned are called si_dict (e.g, s1_dict, s2_dict and s2si_dict). These are: s1_dict = {event_number: s1} (and so on).

4) PMAPS are created by a set of functions in peak_functions_c: namely:
**find_s1**, **find_s2** and  **find_s2si**, which provide a cleaner interface than previous version. The free functions are accessed from the method **pmaps** in base_cities:

def pmaps(self, s1_indx, s2_indx, csum, sipmzs):
        s1 = cpf.find_s1(csum, s1_indx, **self.s1_params._asdict())
        s1 = cpf.correct_s1_ene(s1.s1d, csum)
        s2 = cpf.find_s2(csum, s2_indx, **self.s2_params._asdict())
        s2si = cpf.find_s2si(sipmzs, s2.s2d, thr = self.thr_sipm_s2)
        return s1, s2, s2si

which returns now instances of the high level objects. Analogously, the PMAPS are read back by function **load_pmaps** (now in pmap_io module in io) which in turn call free functions (living in pmaps_functions_c): 

def load_pmaps(PMP_file_name):
    """Read the PMAP file and return transient PMAP rep."""

    s1t, s2t, s2sit = read_pmaps(PMP_file_name)
    s1_dict              = df_to_s1_dict(s1t)
    s2_dict              = df_to_s2_dict(s2t)
    s2si_dict            = df_to_s2si_dict(s2t, s2sit)
    return s1_dict, s2_dict, s2si_dict

The objects returned by **load_pmaps** are si_dict = {event_n0:si}.

5) The high-level classes themselves are created in pmaps.pyx which now lives in a new package called evm (from event model). 

6) The high level pmap classes and all the functions that create and manipulate them are now cython functions, in order to optimize performance and type safety. Previous python functions are kept in some cases (for testing purposes) but removed from the API (eg, by changing their name from name_of_function to _name_of_function). 

7) Tests have been revised and often refactored to describe the changes above. 

8) cities involving pmaps as well as other relevant pieces of code (e.g, s12_selector) have been modified to use the high level PMAP interface rather than the old representation.

jjgomezcadenas committed:
Write sipm_s2_dict and friends in cython.

jjgomezcadenas committed:
Add cython-module ic_types_c to core.

A cython version of ic_types has been added to core. The cython types
are fully compatible with their python cousins but provide
statically-typed (and supposedly faster) objects.

jjgomezcadenas committed:
Create new module evm (event model).

A new module, called evm (for event model) has been created to
centralize the description of the event model.

The high level classes describing PMAPS are the first guests of evm.

jjgomezcadenas committed:
Improved tests in ic_types.

Changed error-prone assert a == b (when using floats) by np
comparison.

jjgomezcadenas committed:
Add find_s12 (temporary fix).

currently function find_S12 returns an old-type objet (a class that
extends a dictionary with a pytables for persistency). Function
find_s12 (notice the difference between s12 and S12) returns simply a
dictionary that can then be passed to a high level S12 interface. The
old function is kept for the time being for backward compatibility.

jjgomezcadenas committed:
Add S2Si high level interface to evm/pmaps

jjgomezcadenas committed:
Cosmetics.

jjgomezcadenas committed:
Added a test on S2Si.

jjgomezcadenas committed:
Cython functions to compute S2Si pmaps.

Cython functions needed to compute S2Si pmaps have been written.

jjgomezcadenas committed:
Re-factor pmaps function in base_cities.

pmaps function has been simplified. In order to comppute pmaps cython
functions returning s1d, s2d and s2sid (dictionary low types) are
invoked. Then, they persistent class is invoked.

jjgomezcadenas committed:
Change from correct_S1_ene to correct_s1_ene

jjgomezcadenas committed:
Function plot_s2si_map takes s2di evm interface.

jjgomezcadenas committed:
Fixed pmap_io_test to use slightly modified interface.

Interface to compute S2Si has changed (one needs to call now
two functions rather than one) and the test has been correspondingly
adapted. Also, the notation has improved (using s1d, s2d, etc. to
name dictionary objects, reserving s1, s2 etc. for high level objects).

jjgomezcadenas committed:
Revised (shorten) doc strings in peak_functions.py

Doc strings were somewhat verbose and have been accordingly shortened.

jjgomezcadenas committed:
Clean up pmaps-plotting functions.

A number of obsolete plotting functions in pmaps_functions have been
removed. In addition, some dead code has been removed from
pmaps_mpl.py

jjgomezcadenas committed:
Refactoring of PMAPS classes.

PMAPS objects have been refactored to enhance the high-level
interface. The refactoring includes:
1) declaration of types inside cython functions (to boost
performance).
2) The dictionaries underlying the S1, S2 and S2Si classes are made
public. Thus, for and s1 object the s1d attribute gives access to the
{peak_no:(t,E) underlying dict.
3) The high level classes own now the method store, and thus are
capable to write themselves.

In summary, **finally** PMAPS are represented by high-level objects.

Tests have been added to demonstrate the new functionality.

jjgomezcadenas committed:
Refactor cython peak_functions.

The cython peak_functions_c module has been refactored to provide a
more consistent API to PMAPS functions. Specifically:

1) find_S12 has been sustituted by find_s1 and find_s2. Notice that s1
and s2 objects are no longer equivalent (they are different classes
derived from s12 which in turn is no longer public).

2) s2si objects are now found by the function sipm_s2sid which takes a
zero-supressed sipm waveform and an s2d object.

3) all functions (with the exception of find_s12 which is kept public
for the time being for compatibility purposes) return high level
objects. find_s1 returns an s1, find_s2 returns an s2 and sipm_s2sid
returns an s2si.

In summary, the cython module provides now all needed functions to
compute pmaps from calibrated sums in PMTs and SiPMs.

jjgomezcadenas committed:
Refactor pmap_io functions.

pmap_io functions now collect pmaps io functions which were previoulsy
scattered in assorted files. Specifically they provide
load_pmaps which returns s1_dict, s2_dict and s2si_dict, where each
one of those dicts is of the form {event_no:s1 (and so on). It also
provides funciton s1_s2_si_from_pmaps which returns the s1, 2, si
objects from the corresponding s1_dict, s2_dict, si_dict, for a given
event.

In turn, the persisten version of S1, S2 and S2Si previously defined
in this module has been removed, classes S1 etc are defined (including
store method) in evm/pmaps.

jjgomezcadenas committed:
event_model has been refactored.

The reco/event_model module has been refactored, removing the
definition of S1, S2, S2Si and Waveform which are now moved to
evm/pmaps (Waveform has been renamed Peak). Tests have been updated
accordingly.

jjgomezcadenas committed:
peak_functions.py has been refactored.

Module peak_functions.py now contains only private functions used for
testing (with one exception to be removed soon).

jjgomezcadenas committed:
pmaps_functions modules have been refactored.

The cython version of pmaps_functions now provide the functionality
needed to load PMAPS from disk: this includes df_to_s1_dict,
df_to_s2_dict and df_to_s2si_dict, which take a data-frame (a
PANDAS persistent table-like representation) of s1, s2, si and return
the corresponding si_dict.

The python version of pmaps currently contain a few functions mostly
related with sipms which will eventually be translated to cython.

jjgomezcadenas committed:
conftest has been adapted to refactored functions and types.

jjgomezcadenas committed:
s1s2_filter has been adapted to refactored functions and types.

jjgomezcadenas committed:
Cities manipulating PMAPS have been adapted to new types.

jjgomezcadenas committed:
Eliminate obsolete function width() in pmaps_functions.py

jjgomezcadenas committed:
Add new functions to S2Sid.

Function sipm_total_energy_dict returns a dictionary {sipm:Q} for each
peak.
Function peak_and_sipm_total_energy_dict(self) returns {n_peak:sipm:Q}

jjgomezcadenas committed:
Function integrate_sipm_charges_in_peak moved to cython functions.

Function now takes an s2si (rather than an s2sid) and a peak
number. Old python function renamed to _integrate_... and kept in
pmaps_functions.py for testing purposes. Tests updated.

jjgomezcadenas committed:
Dorothea takes now the cython function integrate_sipm_charges_

Dorothea takes the cython function integrate_sipme_charges_in_peak
rather than the previous python function.

jjgomezcadenas committed:
Function integrate_sipm_charges_in_peak_as_dict is now private.

The python function integrate_sipm_charges_in_peak_as_dict is now
private and used for testing purposes, functionality in S2Si.

jjgomezcadenas committed:
Function integrate_S2Si_charge is not private.

The python function integrate_S2Si_charge is now private,
functionality now in S2Si class. Tests upgraded.

jjgomezcadenas committed:
Modify s12_filter.py to use s2si rather than python function.

Use s2si.peak_and_sipm_total_energy_dict() rater than python function
integrate_S2Si_charge(s2si.s2sid)

jjgomezcadenas committed:
Refactoring of tests related with pmaps.

Tests related with pmaps have been refactored, introducing new tests
in the high level objects and removing some obsolete tests.

jjgomezcadenas committed:
Delete dead code.

df_pmaps.py was a fossil and has been deleted.

jjgomezcadenas committed:
Use consistent notation in load_pmaps.

Now following the convention: s1_dict = {event_no: s1) and so on.

jmbenlloch committed:
Call peak number peak_no instead of event_no

jmbenlloch commented on invisible_cities/evm/pmaps.pyx:
```diff
@@ -0,0 +1,272 @@
+           s2sid = {peak:{nsipm:[Q]}}
+           Q is the energy in each SiPM sample
+        """
+        S2.__init__(self, s2d)
```
We don't understand why there are two copies of `s2d`, one in the object `s2` and another here in the `s2si`. If we are going to have `s1`, `s2` and `s2si`, this information seems to be redundant.

jjgomezcadenas replied:
It is redundant but nice. Consider the case in which you want to retrieve the waveform of a given SiPM. The underlying s2sid object has only information on E but not on time (since this is own by the s2d object). By keeping a reference to the s2d object one can access the time waveform (and also the energy waveform) of the s2 *directly*, and this is provided by the high level interface without the need for gimmnastics (e.g, go and search for the corresponding information in the s2 object). At the same time, by keeping the reference to s2, we do not need to save the time waveform for each SiPM in the S2Si, thus saving space and speed access.

abotas commented on invisible_cities/cities/base_cities.py:
```diff
@@ -422,17 +414,20 @@ def __init__(self, **kwds):
+    #         t, e, sipms = cpf.rebin_S2(S2[peak][0], S2[peak][1], Si[peak], self.rebin)
+    #         S2_rebin[peak] = Peak(t, e)
+    #         Si_rebin[peak] = sipms
+    #     return S2_rebin, Si_rebin
```
I think we can delete all of this. This code calls a function, `cpf.rebin_S2`, that has not existed in nextic/master for months. `cpf.rebin_S2` seems to be a precursor to the old version of `rebin_waveform` that I recently changed.

jjgomezcadenas replied:
Yes, but there is a point in having the possibility of re-rebinning S2 and S2Si at Dorothea level. Typically you have rebinned the s2 and s2si in bins of 1 mus when creating the pmaps. This bin is often "too thin" (it corresponds to 1 mm) and when creating hits and tracks we will probably increase to 5 mus or even 10 mus. Thus we must be able to rebin by a factor 5 or even 10 at this level. The idea is to call from here @abotas new rebinning functions. Ale, can you do this?

jmbenlloch commented on invisible_cities/cities/dorothea.py:
```diff
@@ -133,9 +144,56 @@ def _create_kr_event(self, evt_number, evt_time, S1, S2, Si):
 
         return evt
+
+    # def _create_kr_event(self, evt_number, evt_time, S1, S2, Si):
```
Is there any reason for this to be here? You have written a new version of the function and the old one is commented.

jjgomezcadenas replied:
Yes, the old version was commented while I was trying to pass tests... now it can be removed.

abotas commented on invisible_cities/cities/penthesilea.py:
```diff
@@ -72,48 +72,58 @@ def _file_loop(self, write_hits):
+
+        s1_t = s1.peak_waveform(0).tpeak
+
+        #S2, Si = self.rebin_s2(S2, Si) TODO REVISE
```
Again, I think this is the same comment: rebin_s2 calls cpf.rebin_S2 which has not existed in our master repository for months.  

If the idea is that penthesilea wants to rebin the s2 and s2si such that they have time bins larger than those saved in the pmaps, we can write a tiny function to do that. It should be very simple since now the `rebin_waveform` function returns clean usually 1 mus time bins. If not, we can just delete this.

jjgomezcadenas replied:
Same comments as before, I need you to include rebinning (new style here)

jmbenlloch commented on invisible_cities/io/pmap_io_test.py:
```diff
@@ -101,144 +174,149 @@ def test_pmap_writer(config_tmpdir, s12_dataframe_converted, s2si_dataframe_conv
-                for nsipm in sipm:
-                    np.testing.assert_allclose(sipm[nsipm], sipm2[nsipm])
+# @mark.slow
+# def test_pmap_electrons_40keV(config_tmpdir):
```
Again, if this is useful shouldn't be commented, if not maybe we should remove it. What do you think @jjgomezcadenas ?

jmbenlloch commented on invisible_cities/reco/event_model.py:
```diff
@@ -54,187 +54,187 @@ def __str__(self):
-        return s + '\n'.join(s2b) + "\n" + "=" * 80
-
-    __repr__ = __str__
+# class Waveform:
```
Why is all this commented?

Commit:
change name of values contained in s2d to s2l instead of S2

S2 is the name of a class

Commit:
initial structure of rebin_s2

Commit:
write more versitile rebin_array function with tests

rebin array can keep or discard remainder, and can rebin by summing or
by taking the mean

Commit:
delete obsolete rebin_array test

Commit:
cythonized rebin_array

Commit:
coordination of rebin_s2si from base_cities to pf

Commit:
add tests for pmap functions that call rebin_array

abotas commented on invisible_cities/evm/pmaps.pyx:
```diff
@@ -0,0 +1,272 @@
+    def __init__(self, s2d):
+        self.s2d = s2d
+        super(S2, self).__init__(s2d)
+
```
One minor issue is that if we want a function `manipulate_si` or `manipulate_s12` that interacts either the dictionary in s1 or s2, we have a small problem because the function does not know whether to use si.s1d or si.s2d, right?

Maybe we use s1.d and s2.d to access their dictionaries, which is also more concise. 

In s2si we can leave things as they are since it has two dictionaries. And, since if a function needs to interact with s2si, it needs to know that it is an s2si since s2sid has a different format than s1d and s2d. What do you think @jjgomezcadenas ?

Commit:
use high level S12 / peak classes to their potential

jmbenlloch committed:
Remove old commented code

jmbenlloch reviewed (approved):
Important changes to add clarity to the pmaps API. This closes #262.
